tensorboard/components/tensor_widget/types.ts

```diff
@@ -0,0 +1,211 @@
+  dtype: string;
+
+  /** Shape of the underlying tensor. */
+  shape: number[];
+
+  /** Get a specific element. */
+  get: (...indices: number[]) => Promise<boolean|number|string>;
+
+  /**
+   * Get a view of the underlying tensor with the specified
+   * slicing and viewing dimensions, as well as the ranges
+   * within the viewing dimensions.
+   */
+  view: (navigation: TensorViewSlicingSpec) => Promise<SlicedValues>;
+
+  /** Get the health pill of the underlying tensor. */
+  getHealthPill: () => Promise<BaseTensorHealthPill>;
+}
+
+/**
+ * Represents the primitive values from slicing a multi-dimensional
+ * tensor.
+ */
+export type SlicedValues =
+    boolean|boolean[]|boolean[][]|boolean[][][]|
+    number|number[]|number[][]|number[][][]|
+    string|string[]|string[][]|string[][][];
+
+/**
+ * A data structure that keeps track of how an n-dimensional array (tensor)
+ * is sliced down to a smaller number of dimensions for visualization
+ * in the tensor widget.
+ *
+ * For example, suppose there is a 4D tensor of shape [16, 128, 128, 3]
+ * representing a NHWC image batch. If you'd like to get the 4-by-3 top-left
+ * corner of the first image of the last color channel  displayed in the tensor
+ * widget, this interface should have the following concrete value:
+ *
+ * ```
+ * {
+ *   slicingDimsAndIndices: [{
+ *     dim: 0,
+ *     index: 0
+ *   }, {
+ *     dim: 3,
+ *     index: 2
+ *   }],
+ *   viewingDims: [1, 2],
+ *   verticalRange: [0, 4],
+ *   horizontalRange: [0, 3]
+ * }
+ * ```
+ */
+export interface TensorViewSlicingSpec {
+  /**
+   * Which dimensions of the tensor are sliced down to a slice of 1.
+   *
+   * - The `dim` field is the 0-based dimension index.
+   * - The `index` is the 0-based index for the selected slice.
+   */
+  slicingDimsAndIndices: Array<{dim: number, index: number}>;
+
+  /**
+   * Which dimensions are used for viewing (i.e., rendered in the
+   * tensor widget, as a table, by default.)
+   *
+   * Possible lengths of this array field:
+   * - 0 for scalar tensors.
+   * - 1 for 1D tensors.
+   * - 2 for 2D+ tensors.
+   */
+  viewingDims: number[];
+
+  /**
+   * The indices from the first viewing dimension, which are shown as rows.
+   *
+   * The two numbers are beginning index (inclusive) and ending index
+   * (exclusive).
+   */
+  verticalRange: [number, number];
+
+  /**
+   * The indices from the second viewing dimension, which are shown as columns.
+   *
+   * The two numbers are beginning index (inclusive) and ending index
+   * (exclusive).
+   */
+  horizontalRange?: [number, number];
+
+  /**
+   * Optional dimension for depth.
+   *
+   * This supports visualization that requires a depth dimension, e.g.,
+   * color channels.
+   */
+  depthDim?: number;
+}
+
+/** Options used during the creation of a single-tensor tensor widget. */
+export interface TensorWidgetOptions {
+  /**
+   * Name of the tenosr.
+   *
+   * Optional. If provided, it will be shown on the widget.
```

Comment:
This is a business-logic of the view that consumes the data. I don't think you want to add that fact in the data type definition.

Reply:
The data type is closely related to the widget's business logic. I can't quite see where this info should go if it isn't here.

Reply:
> I can't quite see where this info should go if it isn't here.

It would go where the state/data is consumed.

Reply:
OK. I'm convinced on this one. Removed it.